### PR TITLE
Bael 7050 Java 21 New Features

### PR DESCRIPTION
**Pull request by musibs**

### CODE REVIEW

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java [hunk 1]
```diff
@@ -0,0 +1,42 @@
+package com.baeldung.java21;
+
+public class PatternCaseLabels {
+
+    static String processInputOld(String input) {
+        String output = null;
```
Comment (JavierPeris) [inline core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java:6]: There's no need to initialize the variable. It's redundant.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+            when "Yes".equalsIgnoreCase(s) -> {
+                output = "It's Yes";
+            }
+            case String s
+            when "No".equalsIgnoreCase(s) -> {
+                output = "It's No";
+            }
+            case String s -> {
+                output = "Try Again";
+            }
```
Comment (JavierPeris) [inline core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java:38]: We don't need code blocks, we can simplify:
```suggestion
            when "Yes".equalsIgnoreCase(s) -> output = "It's Yes";
            case String s
            when "No".equalsIgnoreCase(s) -> output = "It's No";
            case String _ -> output = "Try Again";
```

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/StringTemplates.java [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package com.baeldung.java21;
+
+public class StringTemplates {
+
+    public String getStringTemplate() {
+        String name = "Baeldung"; 
+        String welcomeText = STR."Welcome to \{name}"; 
+        return welcomeText;
```
Comment (JavierPeris) [inline core-java-modules/core-java-21/src/main/java/com/baeldung/java21/StringTemplates.java:8]: We can simplify to:
```suggestion
        return STR."Welcome to \{name}";
```

Anyway, this is a feature in preview, isn't it?

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+            balance = ta.getBalance();
+        }
+        else if(account instanceof CurrentAccount ca) {
+            balance = ca.getBalance();
+        }
+        return balance;
+    }
+    
+    static double getBalanceWithSwitchPattern(Account account) {
+        double result = 0;
```
Comment (JavierPeris) [inline core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:42]: There's no need to initialize the variable, it's redundant.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        switch (account) {
+            case null -> throw new RuntimeException("Oops, account is null");
+            case SavingsAccount sa -> result = sa.getBalance();
+            case TermAccount ta -> result = ta.getBalance();
+            case CurrentAccount ca -> result = ca.getBalance();
+            default -> result = account.getBalance();
+        };
```
Comment (JavierPeris) [inline core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:49]:
```suggestion
        double result;
        switch (account) {
            case null -> throw new IllegalArgumentException("Oops, account is null");
            case SavingsAccount sa -> result = sa.getBalance();
            case TermAccount ta -> result = ta.getBalance();
            case CurrentAccount ca -> result = ca.getBalance();
            default -> result = account.getBalance();
        }
```

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    class TermAccount extends Account {
+        double getBalance() {
+            return 1000;
+        } 
+    }
+    class CurrentAccount extends Account {
+        double getBalance() {
+            return 10000;
+        } 
+    }
```
Comment (JavierPeris) [inline core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:25]:
```suggestion
    static class Account{
        double getBalance() {
            return 0;
        }
    }
    
    static class SavingsAccount extends Account {
        @Override
        double getBalance() {
            return 100;
        } 
    }
    static class TermAccount extends Account {
        @Override
        double getBalance() {
            return 1000;
        } 
    }
    static class CurrentAccount extends Account {
        @Override
        double getBalance() {
            return 10000;
        } 
    }
```

**Review by JavierPeris — CHANGES_REQUESTED** [review]
Please use JUnit 5 for tests.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java (hunk 1 above)
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java:6]: Removed the variable initialization to null.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java (hunk 2 above)
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java:38]: Removed the code blocks and moved to single-line statements.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/StringTemplates.java (hunk 3 above)
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/StringTemplates.java:8]: Yes, it is a preview feature. I have simplified the return statement.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java (hunk 4 above)
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:42]: Removed the variable initialization to null.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+        else if(account instanceof CurrentAccount ca) {
+            balance = ca.getBalance();
+        }
+        return balance;
+    }
+    
+    static double getBalanceWithSwitchPattern(Account account) {
+        double result = 0;
+        switch (account) {
+            case null -> throw new RuntimeException("Oops, account is null");
```
[the remark this answers is not in the document]
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:44]: Moved to use IllegalArgumentException.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java (hunk 5 above)
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:49]: Moved to use IllegalArgumentException.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java (hunk 6 above)
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:25]: Moved to static class type and test case is also updated.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+        }
+        }
+        return output;
+    }
+
+    static String processInputNew(String input) {
+        String output;
+        switch (input) {
```
[the remark this answers is not in the document]
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java:24]: The default case is not allowed. It throws the following compilation error:

Switch case cannot have both unconditional pattern and default label

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+    
+    static double getBalanceWithSwitchPattern(Account account) {
+        double result;
+        switch (account) {
+            case null -> throw new IllegalArgumentException("Oops, account is null");
+            case SavingsAccount sa -> result = sa.getBalance();
+            case TermAccount ta -> result = ta.getBalance();
+            case CurrentAccount ca -> result = ca.getBalance();
+            default -> result = account.getBalance();
+        };
```
[the remark this answers is not in the document]
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:50]: Removed.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+            return 100;
+        } 
+    }
+    
+    static class TermAccount extends Account {
+        double getBalance() {
+            return 1000;
+        } 
+    }
+    class CurrentAccount extends Account {
```
[the remark this answers is not in the document]
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:22]: Changed to static.

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java [hunk 11]
```diff
@@ -0,0 +1,53 @@
...
+    static class SavingsAccount extends Account {
+        double getBalance() {
+            return 100;
+        } 
+    }
+    
+    static class TermAccount extends Account {
+        double getBalance() {
+            return 1000;
+        }
```
[the remark this answers is not in the document]
Comment (musibs) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/SwitchPattern.java:20]: Added @Override annotation for the overridden methods.

File: core-java-modules/core-java-21/src/test/java/com/baeldung/java21/PatternCaseLabelsUnitTest.java [hunk 12]
```diff
@@ -0,0 +1,48 @@
+package com.baeldung.java21;
+
+import org.junit.Assert;
+import org.junit.Test;
```
[the remark this answers is not in the document]
Comment (musibs) [reply core-java-modules/core-java-21/src/test/java/com/baeldung/java21/PatternCaseLabelsUnitTest.java]: Updated to JUnit 5

File: core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (JavierPeris) [reply core-java-modules/core-java-21/src/main/java/com/baeldung/java21/PatternCaseLabels.java:24]: Thanks for letting me know!

File: core-java-modules/core-java-21/src/test/java/com/baeldung/java21/PatternCaseLabelsUnitTest.java [hunk 13]
```diff
@@ -0,0 +1,48 @@
+package com.baeldung.java21;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import org.junit.jupiter.api.Test;
+
+public class PatternCaseLabelsUnitTest {
+
+    @Test
+    public void whenProcessInputOldWayWithYes_thenReturnOutput() {
```
Comment (JavierPeris) [inline core-java-modules/core-java-21/src/test/java/com/baeldung/java21/PatternCaseLabelsUnitTest.java:9]: Using JUnit 5 we don't need to make class nor method public.

Comment (musibs) [reply core-java-modules/core-java-21/src/test/java/com/baeldung/java21/PatternCaseLabelsUnitTest.java]: Done

**Pull request merged**